### PR TITLE
build: Bump minimum required clang/libc++ to 8.0

### PR DESCRIPTION
**Pull request by maflcko**
This is not for 23.0, but for 24.0. It comes with the following benefits:

* Can use C++17 P0083R3 std::set::merge from libc++ 8.0
* No longer need to provide support for clang-7, which already fails to compile on some architectures (https://github.com/bitcoin/bitcoin/issues/21294#issuecomment-998098483)

This should be fine, given that all supported operating systems ship with at least clang-10:

* CentOS 8: clang-12
* Stretch: https://packages.debian.org/stretch/clang-11
* Buster: https://packages.debian.org/buster-backports/clang-11
* Bionic: https://packages.ubuntu.com/bionic-updates/clang-10
* Focal: https://packages.ubuntu.com/focal/clang-10

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

**Comment by shaavan** [comment]
Concept ACK

On a side note, I was wondering about one thing. Would this PR remain unmerged until v23.0 finally releases on April 1st?

**Comment by hebasto** [comment]
> Would this PR remain unmerged until v23.0 finally releases on April 1st?

Until branching 23.x off, I guess.

**Comment by MarcoFalke** [comment]
Rebased to get fresh CI.

Fun fact: I couldn't compile current master with clang-7 and system libs on Bionic, so this could theoretically even be merged before 23.0

**Comment by hebasto** [comment]
@MarcoFalke 
> Fun fact: I couldn't compile current master with clang-7 and system libs on Bionic, so this could theoretically even be merged before 23.0

Is it similar to https://github.com/bitcoin/bitcoin/issues/21294#issuecomment-998098483?

**Comment by MarcoFalke** [comment]
Rebased

**Comment by MarcoFalke** [comment]
Rebased

**Review by fanquake — APPROVED** [review]
ACK fae20e6b50306f91c74037e915aa0ab75a0a6b3b - I think this is fine to do. I would be surprised if in another 6 months time someone was stuck on a system we supported, needing to compile Core, and only had access to Clang 7 or older. As mentioned in the PR description, all systems we currently support, already support multiple newer versions of Clang.

File: doc/dependencies.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ These are the dependencies currently used by Bitcoin Core. You can find instruct
 | --- | --- | --- | --- | --- | --- |
 | Berkeley DB | [4.8.30](https://www.oracle.com/technetwork/database/database-technologies/berkeleydb/downloads/index.html) | 4.8.x | No |  |  |
 | Boost | [1.77.0](https://www.boost.org/users/download/) | [1.64.0](https://github.com/bitcoin/bitcoin/pull/22320) | No |  |  |
-| Clang<sup>[ \* ](#note1)</sup> |  | [7.0](https://releases.llvm.org/download.html) (C++17 & std::filesystem support) |  |  |  |
+| Clang |  | [8.0](https://releases.llvm.org/download.html) (C++17 & std::filesystem support) |  |  |  |
```
Comment (hebasto) [inline doc/dependencies.md:10]:
```suggestion
| Clang |  | [8.0](https://releases.llvm.org/download.html) (C++17, std::filesystem, and P0083R3 support) |  |  |  |
```

**Review by hebasto — APPROVED** [review]
ACK fae20e6b50306f91c74037e915aa0ab75a0a6b3b, I have reviewed the code and it looks OK, I agree it can be merged.

File: doc/dependencies.md (hunk 1 above)
Comment (MarcoFalke) [reply doc/dependencies.md:10]: I don't think mentioning a paper in the dependencies will give users any useful information. If they are interested in this, they can use https://en.cppreference.com/w/cpp/compiler_support or the documentation of their compiler. Practically they should only care about the minimum required version.

**Pull request merged**